### PR TITLE
fix bug PrintPreviewControl lacks UI Automation Scroll pattern support

### LINKED ISSUE #9584: PrintPreviewControl lacks UI Automation Scroll pattern support
### .NET version

.NET SDK 8.0.0-rc.1.23371.3

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

Yes, regression after PR #9477 

### Issue description

After replacing native scrolling with .NET ScrollBars in PR #9477 PrintPreviewControl no longer supports Scroll pattern.

**Before PR #9477**
Narrator was announcing scroll percent of PrintPreviewControl:

PrintPreviewControl_Narrator_before_PR_9477

ScrollPattern was supported:
PrintPreviewControl_before_PR_9477

**After PR #9477**
Narrator does not announce scroll percent:
PrintPreviewControl_Narrator_after_PR_9477

ScrollPattern is not supported:
PrintPreviewControl_after_PR_9477

### Steps to reproduce

1. Create .NET project with PrintPreviewControl and PrintDocument, or run Accessibility_Core_App.
2. Run Narrator and switch focus to PrintPreviewControl.

### PR DESCRIPTION
**Pull request by Epica3055**
Fixes #9584

## Proposed changes

- 
- Add scroll pattern support
- 

## Regression? 

- Yes 

## Risk

- Minimal 

### Before

PrintPreviewControl_after_PR_9477

### After

PrintPreviewControl_before_PR_9477

## Test methodology 

- 
-  Manually 
- 

 

## Test environment(s) 

.Net 8.0.0-rc.1.23404.1

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9694)

### CODE REVIEW

**Comment by elachlan** [comment]
@JeremyKuhne you might want to check the interior stuff here.

**Comment by Tanya-Solyanik** [comment]
@Epica3055  - could you please define IScrollProvider using CSWin32, the same way as IENUMVariant is defined. First add the name of the interface to https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms.Primitives/src/NativeMethods.txt
and

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.cs [hunk 1]
```diff
@@ -444,6 +444,28 @@ private Point Position
...
+    }
+
+    /// <summary>
+    /// Equal to the horizontal percentage of the entire control that is currently viewable.
+    /// </summary>
+    internal double HorizontalViewSize
+    {
+        get
+        {
+            return InsideRectangle.Width * 100.0 / VirtualSize.Width;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.cs:465]: Please use expression-bodied properties in this file.

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.cs [hunk 2]
```diff
@@ -444,6 +444,28 @@ private Point Position
         }
     }
 
+    /// <summary>
+    /// Equal to the vertical percentage of the entire control that is currently viewable.
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.cs:448]: Please indent doc comments by 2 spaces in all new code you write, we are trying to adopt this stype.

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs [hunk 3]
```diff
@@ -78,5 +86,47 @@ internal override UiaCore.IRawElementProviderFragmentRoot FragmentRoot
                     return base.FragmentNavigate(direction);
             }
         }
+
+        public void Scroll(ScrollAmount horizontalAmount, ScrollAmount verticalAmount) { }
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs:90]: Why is this method and the next not implemented?

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs [hunk 4]
```diff
@@ -52,6 +52,14 @@ public override int GetChildCount()
             };
         }
 
+        internal override bool IsPatternSupported(UiaCore.UIA patternId)
+            => patternId switch
+            {
+                UiaCore.UIA.ScrollPatternId => this.TryGetOwnerAs(out PrintPreviewControl? owner)
+                    && (owner._vScrollBar.Visible|| owner._hScrollBar.Visible) ? true : false,
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs:59]:
```suggestion
                    && (owner._vScrollBar.Visible || owner._hScrollBar.Visible) ? true : false,
```

**Review by Tanya-Solyanik — COMMENTED** [review]
Implementation looks good. Please define the interface using CSwin32 and add a unit test.

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs [hunk 5]
```diff
@@ -78,5 +88,124 @@ internal override UiaCore.IRawElementProviderFragmentRoot FragmentRoot
...
+        }
+
+        public double HorizontalViewSize => this.TryGetOwnerAs(out PrintPreviewControl? owner)
+            && owner._hScrollBar.Visible ? owner.HorizontalViewSize : 100;
+
+        public double VerticalViewSize => this.TryGetOwnerAs(out PrintPreviewControl? owner)
+            && owner._vScrollBar.Visible ? owner.VerticalViewSize : 100;
+
+        BOOL IScrollProvider.Interface.HorizontallyScrollable =>
+            this.TryGetOwnerAs(out PrintPreviewControl? owner) ? (BOOL)owner._hScrollBar.Visible : (BOOL)false;
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs:206]: nit: I don't think the casting is necessary here.
```suggestion
            this.TryGetOwnerAs(out PrintPreviewControl? owner) ? owner._hScrollBar.Visible : false;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs [hunk 6]
```diff
@@ -78,5 +88,124 @@ internal override UiaCore.IRawElementProviderFragmentRoot FragmentRoot
...
+
+        BOOL IScrollProvider.Interface.VerticallyScrollable =>
+            this.TryGetOwnerAs(out PrintPreviewControl? owner) ? (BOOL)owner._vScrollBar.Visible : (BOOL)false;
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs:209]: Ditto here.
```suggestion
            this.TryGetOwnerAs(out PrintPreviewControl? owner) ? owner._vScrollBar.Visible : false;
```

**Comment by Tanya-Solyanik** [comment]
@Epica3055 - please rebase for Main because this is a regression

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs [hunk 7]
```diff
@@ -78,5 +88,135 @@ internal override UiaCore.IRawElementProviderFragmentRoot FragmentRoot
...
+                    case ScrollAmount.ScrollAmount_NoAmount:
+                        return HRESULT.E_FAIL;
+                }
+
+                return HRESULT.S_OK;
+            }
+
+            if (owner._vScrollBar.Visible && horizontalAmount != ScrollAmount.ScrollAmount_NoAmount)
+            {
+                switch (horizontalAmount)
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs:129]: Should this be the following?
```suggestion
            if (owner._vScrollBar.Visible && verticalAmount!= ScrollAmount.ScrollAmount_NoAmount)
            {
                switch (verticalAmount)
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs [hunk 8]
```diff
@@ -78,5 +88,135 @@ internal override UiaCore.IRawElementProviderFragmentRoot FragmentRoot
...
+                    case ScrollAmount.ScrollAmount_LargeDecrement:
+                        scrollValue = owner._vScrollBar.Value - owner._vScrollBar.LargeChange;
+                        owner._vScrollBar.Value = (scrollValue < owner._vScrollBar.Minimum ? owner._vScrollBar.Minimum : scrollValue);
+                        break;
+                    case ScrollAmount.ScrollAmount_SmallDecrement:
+                        scrollValue = owner._vScrollBar.Value - owner._vScrollBar.SmallChange;
+                        owner._vScrollBar.Value = (scrollValue < owner._vScrollBar.Minimum ? owner._vScrollBar.Minimum : scrollValue);
+                        break;
+                    case ScrollAmount.ScrollAmount_NoAmount:
+                        return HRESULT.E_FAIL;
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs:148]: Same here. This case will never get hit due to the if condition

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs [hunk 9]
```diff
@@ -52,6 +54,14 @@ public override int GetChildCount()
             };
         }
 
+        internal override bool IsPatternSupported(UiaCore.UIA patternId)
+            => patternId switch
+            {
+                UiaCore.UIA.ScrollPatternId => this.TryGetOwnerAs(out PrintPreviewControl? owner)
+                    && (owner._vScrollBar.Visible || owner._hScrollBar.Visible) ? true : false,
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs:61]:
```suggestion
                UiaCore.UIA.ScrollPatternId => this.TryGetOwnerAs(out PrintPreviewControl? owner)
                    && (owner._vScrollBar.Visible || owner._hScrollBar.Visible),
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs [hunk 10]
```diff
@@ -78,5 +88,125 @@ internal override UiaCore.IRawElementProviderFragmentRoot FragmentRoot
...
+        }
+
+        public double HorizontalViewSize => this.TryGetOwnerAs(out PrintPreviewControl? owner)
+            && owner._hScrollBar.Visible ? owner.HorizontalViewSize : 100;
+
+        public double VerticalViewSize => this.TryGetOwnerAs(out PrintPreviewControl? owner)
+            && owner._vScrollBar.Visible ? owner.VerticalViewSize : 100;
+
+        BOOL IScrollProvider.Interface.HorizontallyScrollable =>
+            this.TryGetOwnerAs(out PrintPreviewControl? owner) ? owner._hScrollBar.Visible : false;
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs:207]:
```suggestion
            this.TryGetOwnerAs(out PrintPreviewControl? owner) && owner._hScrollBar.Visible;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs [hunk 11]
```diff
@@ -78,5 +88,125 @@ internal override UiaCore.IRawElementProviderFragmentRoot FragmentRoot
...
+
+        BOOL IScrollProvider.Interface.VerticallyScrollable =>
+            this.TryGetOwnerAs(out PrintPreviewControl? owner) ? owner._vScrollBar.Visible : false;
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibleObject.cs:210]:
```suggestion
            this.TryGetOwnerAs(out PrintPreviewControl? owner) && owner._vScrollBar.Visible;
```

**Review by lonitra — APPROVED** [review]
Some small suggestions around bools, but otherwise looks good 👍

**Review by lonitra — DISMISSED** [review]

**Comment by lonitra** [comment]
@Tanya-Solyanik I am assuming this needs to be ported to release/8.0 branch as well?

**Review by SimonZhao888 — APPROVED** [review]

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibilityObjectTests.cs [hunk 12]
```diff
@@ -138,6 +140,41 @@ public void PrintPreviewControlAccessibilityObject_FragmentNavigate_ReturnsConsi
         Assert.Equal(accessibleObject, horizontalScrollBarAO.FragmentNavigate(Interop.UiaCore.NavigateDirection.Parent));
     }
 
+    [WinFormsFact]
+    public void PrintPreviewControlAccessibilityObject_ScrollInfo_ReturnsConsistent()
+    {
+        using SubPrintPreviewControl control = new();
+
+        PrintPreviewControlAccessibleObject accessibleObject = (PrintPreviewControlAccessibleObject)control.AccessibilityObject;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Printing/PrintPreviewControl.PrintPreviewControlAccessibilityObjectTests.cs:148]:
```suggestion
        var accessibleObject = (PrintPreviewControlAccessibleObject)control.AccessibilityObject;
```

**Review by Tanya-Solyanik — COMMENTED** [review]
Looks good. Please create another PR into release/8.0 branch, this was a regression so we want to fix it in the release where it regressed, and there is not automatic flow between these branches.

**Review by Tanya-Solyanik — APPROVED** [review]

**Pull request merged**